### ISSUE TITLE
failed contact linker attempt

### ISSUE THREAD

**Issue by bradfordcondon** [tripal/tripal#217]
We get the below error trying to create an arraydesign:

```Notice: Undefined index: arraydesign in chado_linker__contact_widget->form() (line 27 of /Users/chet/UTK/tripal/sites/all/modules/tripal/tripal_chado/includes/TripalFields/chado_linker__contact/chado_linker__contact_widget.inc).
Warning: key() expects parameter 1 to be array, null given in chado_linker__contact_widget->form() (line 27 of /Users/chet/UTK/tripal/sites/all/modules/tripal/tripal_chado/includes/TripalFields/chado_linker__contact/chado_linker__contact_widget.inc).
Notice: Undefined index: arraydesign in chado_linker__contact_widget->form() (line 28 of /Users/chet/UTK/tripal/sites/all/modules/tripal/tripal_chado/includes/TripalFields/chado_linker__contact/chado_linker__contact_widget.inc).
```

arraydesign has a contact foreign key (manufacturer). I already made a field for it.  I assume thats why its attaching?

But... We have no arraydesign_contact linker table.

**Comment by spficklin** [comment]
Can you clarify what you mean when you said you "made a field for it"?  Do you mean you made a field for the contact_id field for the arraydesign table?  If so, is it generic enough to support any table that may have a 'contact_id' column?

**Comment by bradfordcondon** [comment]
>Do you mean you made a field for the contact_id field for the arraydesign table? If so, is it generic enough to support any table that may have a 'contact_id' column?

Yes.  Note that the column is manufacturer_id with the contact FK.  I thought I would need a new field because of this so I made one: [obi__manufacturer](https://github.com/tripal/tripal_analysis_expression/tree/master/tripal_protocol/includes/TripalFields/obi__manufacturer).  I did not write it with "generic field" in mind unfortunately

This field doesnt invoke chado_linker__contact, however.  My theory is that chado_linker__contact attaches to any table with a contact FK?  If thats the case we can write in another exception for biomaterials in this field instead?

**Comment by laceysanderson** [comment]
Hmmm... chado_linker__contact only attaches when [tablename]_contact exists. In this case arraydesign_contact doesn't exist so it shouldn't be attached. See https://github.com/tripal/tripal/blob/7.x-3.x/tripal_chado/includes/tripal_chado.fields.inc#L470. Can you confirm this is the case by looking at the "Manage Fields" display?

**Comment by bradfordcondon** [comment]
Sorry for not getting back on this.


<img width="1409" alt="screen shot 2018-01-14 at 9 01 56 pm" src="https://user-images.githubusercontent.com/7063154/34923940-34e21c6c-f96e-11e7-9535-d8758bd74cae.png">

editing the field i see `chado_linker__contact` so i do think this field does attach?

I do not have an `array_design_contact` table.

**Comment by spficklin** [comment]
Okay, this was a comedy of errors.... Tripal actually has two contact fields. One supports the [base_table]_contact linker tables and another for supporting fields that have an FK constraint directly to the contact table within the base table.   The chado_linker__contact is for the former and the local__contact is for the latter.  But we have made a mistake in that both field use the exact same term (local:contact).  We should never have fields that use the same term.  So to fix this we should

1) Add a check to Tripal that warns if a field uses the same term as another
2) Find a solution so that we don't have two fields with the same term. We could:
2a)  Unify the two into single field that handles both cases... that would get too confusing I think
2b)  Remove the local__conact field and use specific fields like @bradfordcondon has done. I can't think of any base tables that use a direct link to the contact table that doesn't have a specific use. 
2c)  Keep both fields but find different terms for them  (may be too problematic)

My favorite solution is 2b. Let's dump the local__contact field and, @bradfordcondon, if your manufacturer field is working we can pull that into the tripal_chado module.

**Comment by bradfordcondon** [comment]
Thanks @spficklin 

What about other fields that require both a FK and linker table?  contacts not the only one.  Will we always propose to use seperate terms? 
 
Manufacturer is[ hardcored for biomaterials right now](https://github.com/tripal/tripal_analysis_expression/tree/master/tripal_protocol/includes/TripalFields/obi__manufacturer) and is missing some calls you've requested elsewhere.  I'll add updating it to my to-do list.

**Comment by spficklin** [comment]
Okay, so here's my suggestion.  We keep the local__contact field, but rather than leave it with the default local:contact term, we programmatically adjust it when we add fields for non linker tables. So we need a CV term for each one.  This way we only need to have one field to handle all of these table columns but each can have it's own settings.

Here is a list of suggestions.  Any objections/changes?

| Table | Column | Term ID | Term |
| ------------- | ------------- | ------------- | ------------- |
| stockcollection | contact_id  | [NCIT:C47954](http://purl.obolibrary.org/obo/NCIT_C47954) |Communication contact |
| arraydesign | manufacturer_id  | [NCIT:C25392](http://purl.obolibrary.org/obo/NCIT_C25392) | Manufacturer |
| assay | operator_id | [NCIT:C48036](http://purl.obolibrary.org/obo/NCIT_C48036) | Operator |
| biomaterial | biosourceprovider_id | [NCIT:C37900](http://purl.obolibrary.org/obo/NCIT_C37900) | Provider |
| quantification | operator_id | [NCIT:C48036](http://purl.obolibrary.org/obo/NCIT_C48036) | Operator |
| study | contact_id | [NCIT:C127533](http://purl.obolibrary.org/obo/NCIT_C127533) | Primary Study Contact |

Having all of these from the same vocabulary makes it slightly easier to work with.

**Comment by bradfordcondon** [comment]
I like that solution.  
I like the terms.    communication contact doesn't quite fit:   "A channel for communication between groups".  The field here is for the lab or group who houses a stock collection... how about [owner](https://www.ebi.ac.uk/ols/ontologies/ncit/terms?iri=http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FNCIT_C42704).  Or [Collected by](https://www.ebi.ac.uk/ols/ontologies/ncit/terms?iri=http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FNCIT_C45262)

They are all NCIT which I have been avoiding after our previous discussion.  This is a case where I can't imagine we'll find accurate terms for all the different types of contacts anywhere else.  Does this mean your opinion on using NCIT in core has changed?  see #247

**Comment by spficklin** [comment]
Okay, I'm good with Collected By.  That makes sense.

NCIT provided the best terms for all of these. In general, I try to pick ontologies that make sense for the data being stored.   Most data being stored in Tripal sites is not cancer specific.  But I couldn't find other good alternatives.

**Comment by bradfordcondon** [comment]
@spficklin what do we do in other situations where there is need for both a linker table field and a linker column field?

On my site we use an organism_analysis linker table.  We wrote a really wonky field to query it...  but i realize that a generic organism linker table field would be good.  But we already have the obi__organism linker column field.

I'm looking at chado and trying to predict where else this problem might pop up.. some things in mage look suspect (assay shows up as a linker in study_assay and a column in acqusition for example, but mage is so confusing in general...)

Is teh general rule of thumb to try and use more specific terms for the case which is more the exception?  It's a bit backwards for organism: in that case, organism is typically a foreign key column in a base table instead of a linker table...

**Comment by bradfordcondon** [comment]
Analysis is another case.  It's both a foreign key (phylotree, quantification) and sometimes a linker table (analysisfeature?  project_analysis... analysis_pub)

**Comment by spficklin** [comment]
Actually, now that I think on it a bit more, I don't think we're going to have any conflicts even if the two fields use the same term.  I can't think of any examples where the base table has a contact_id, organism_id, or analysis_id and also has a linker table as well to any of those tables.  So, this means that even though the two contact fields currently use the same term they'll never both be attached at the same time to a content type.  Although I do like the terms we selected above for the different contacts foreign keys and I think we should still implement that.  

In the case of your organism linker field I think that's the right way to go and even if you use the same term (OBI:organism) I don't think there will ever be any conflicts with the obi__organism field because it shouldn't ever be attached to the same content type as your custom linker table field.  And it makes sense for it to be OBI:organism because that's what it represents.

**Comment by laceysanderson** [comment]
Marking this as a task rather then a bug then? Or do we still want to implement it ASAP (before release as indicated by the tag)?

**Comment by spficklin** [comment]
I think this issue is resolved with PR #398. When that gets merged this can be closed.

**Comment by laceysanderson** [comment]
PR Merged.
